README.md
Yandex.Tank API
===============

This is an HTTP server that controls Yandex.Tank execution. It allows the client to:
* set a breakpoint before an arbitrary test stage (and reset it later)
* launch Yandex.Tank and begin the test
* upload files into the test working directory
* terminate the launched Tank at an arbitrary moment
* obtain the test status
* download the artifacts in a safe manner, without interference to another tests

API-managed Tank
------------------
General information on Yandex.Tank installation and usage can be found in its [documentation](http://yandextank.readthedocs.org).
This section covers the difference between console Tank and API-managed Tank configuration and provides more details on the sequence of the test stages.

### Tank configuration
API-managed Tank is configured via configuration files only. They have the same syntax and options as the console Tank configs.

The configuration parameters are applied in the following order:
  1. common Yandex.Tank configuration files that reside in `/etc/yandex-tank/`
  2. Yandex.Tank API defaults in `/etc/yandex-tank-api/defaults`
  3. **the configuration file sent by the client when launching the test**
  4. Yandex.Tank API overriding configs in `/etc/yandex-tank-api/override`

### Test stages

When the client launches a new test, a new *session* is created and a separate *Tank worker* process is spawned. After this, the test stages are executed in the following order:
  1. **lock**

     Attempt to acquire the tank lock. This stage fails if another test started via console is running.

  2. **init**

     Logging is set up and tank configuration files are read at this stage.

  3. **configure**

     The *configure()* method is called for each module. Most of the dynamic configuration is done here.

  4. **prepare**

     The *prepare_test()* method is called for each module. Heavy and time-consuming tasks (such as stpd-file generation and monitoring agent setup) are done here.

  5. **start**

     The *start_test()* method is called for each module. This should take as little time as possible. Load generation begins at this moment.

  6. **poll**

     Once per second the *is_test_finished()* method is called for each module. This stage ends when any of the modules requests to stop the test.

  7. **end**

     The *end_test()* method is called for each module. This should take as little time as possible.

  8. **postprocess**

     The *post_process()* method is called for each module. Heavy and time-consuming tasks are performed here.

  9. **unlock**

     The tank lock is released and the Tank worker exits.

  10. **finished**

     This is a virtual stage. Reaching this stage means that the Tank worker has already terminated.

The last session status is temporarily stored after tank exit.
The test artifacts are stored forever and should be deleted by external means when not needed.

### Pausing the test sequence
When the session is started, the client can specify the test stage before which the test will be paused (the breakpoint) .
After completing the stages preceding the breakpoint, the Tank will wait until the breakpoint is moved further. You cannot move the breakpoint back.

The breakpoint can be set *before* any stage. One of the most frequent use cases is to set the breakpoint before the **start** stage to synchronize several tanks.
Another is setting the breakpoint before the unlock stage to download the artifacts without interference to other tests.
Third is setting the breakpoint before the init stage to upload additional files.
Beware that setting the breakpoint between the **init** and the **poll** stages can lead to very exotic behaviour.

API requests
-----------

All API requests are asynchronous: we do not wait the tank process to perform requested action.
HTTP code 200 is returned when no error occured while processing the request.
However, this does not necessarily mean that the requested action will be successfully performed.
The client should check the session status to detect Tank failures.

All handles, except for /artifact, return JSON. On errors this is a JSON object with a key 'reason'.

### List of API requests
1. **POST /run?[test=...]&[break=...]**

  Request body: Yandex.Tank config in .ini format (the same as for console Tank)

  Creates a new session with an unique *session ID* and launches a new Tank worker.

  Parameters:
    * test: Prefix of the session ID. Should be a valid directory name. *Default: current datetime in the %Y%m%d%H%M%S format*
    * break: the test stage before which the tank will stop and wait until the next break is set. *Default: "finished"*

  Reply on success:     
  ```json
  {
    "session": "20150625210015_0000000001", //ID of the launched session
    "test": "20150625210015_0000000001" //Deprecated, do not use
  }
  ```

  Error codes and corresponding reasons in the reply:
    * 400, 'Specified break is not a valid test stage name.'
    * 409, 'The test with this ID is already running.'
    * 409, 'The test with this ID has already finished.'
    * 503, 'Another session is already running.'

2. **GET /run?session=...&[break=...]**

  Sets a new break point for the running session.

  Parameters:
    * session: session ID
    * break: the test stage before which the tank will stop and wait until the next break is set. *Default: "finished"*

  Return codes and corresponding reasons:
    * 200, 'Will try to set break before [new break point]'
    * 400, 'Specified break is not a valid test stage name.'
    * 404, 'No session with this ID.'
    * 418, ... (returned when client tries to move the break point back)
    * 500, 'Session failed.'

3. **GET /stop?session=...**

  Terminates the current test.

  Parameters:
    * session: ID of the session to terminate

  Return codes and corresponding reasons:
    * 200, 'Will try to stop tank process.'
    * 404, 'No session with this ID.'
    * 409, 'This session is already stopped.'

4. **GET /status?session=...**

  Returns the status of the specified session.
  Parameters:
    * session: ID of the session.

  Status examples:
  ```json
  {
    "status": "running",
    "stage_completed": true,
    "break": "start",
    "current_stage": "prepare",
    "test": "9f43f3104c2549b98bf74b817dc71cef",
    "failures": []
  }
  ```

  ```json
  {
    "status": "failed", 
    "retcode": 1, 
    "stage_completed": true, 
    "break": "finished", 
    "current_stage": "finished", 
    "test": "9f43f3104c2549b98bf74b817dc71cef", 
    "failures": [
        {
            "reason": "Interrupted", 
            "stage": "prepare"
        }
    ]
  }
  ```

  Error code and the corresponding reason:
    * 404, 'No session with this ID.'

5. **GET /status?**

  Returns a JSON object where keys are known session IDs and values are the corresponding statuses.

6. **GET /artifact?session=...**

  Returns a JSON array of artifact filenames.

  Parameters:
    * test: ID of the test

  Error codes and the corresponding reasons:
    * 404, 'No test with this ID found.'
    * 404, 'Test was not performed, no artifacts.'

7. **GET /artifact?session=...&filename=...**

  Sends the specified artifact file to the client.

  Parameters:
    * session: ID of the session
    * filename: the artifact file name

  Error codes and the corresponding reasons:
    * 404, 'No session with this ID found'
    * 404, 'Test was not performed, no artifacts.'
    * 404, 'No such file'
    * 503, 'File is too large and test is running' (when the file size exceeds 128 kB and some test is running)

8. **POST /upload?session=...&filename=...**

  Stores the request body on the server in the tank working directory for the session under the specified filename.
  The session should be running.

  Parameters:
    * session: ID of the session
    * filename: the name to store the file under

  Error codes and the corresponding reasons:
    * 404, 'Specified session is not running'

### Writing plugins

Some custom plugins might need to know if they are wokring in the console Tank or under API.

API worker process uses a subclass of a standard TankCore as a tank core.
Thus, a plugin can detect execution under API by simply checking that
```python
str(self.core.__class__)=='yandex_tank_api.worker.TankCore'
```




debian/changelog
yandex-tank-api (0.0.11) trusty; urgency=low

  * build for launchpad

 -- Alexey Lavrenuke (load testing) <direvius@yandex-team.ru>  Thu, 25 Jun 2015 15:48:21 +0300

yandex-tank-api (0.0.6-1) unstable; urgency=low

  * source package automatically created by stdeb 0.6.0+git

 -- Andrei Sekretenko <asekretenko@gmail.com>  Wed, 01 Apr 2015 14:39:21 +0300


yandex_tank_api/webserver.py
#!/usr/bin/env python
"""
Yandex.Tank HTTP API: request handling code
"""

import tornado.ioloop
import tornado.web
try:
    from pyjade.ext.tornado import patch_tornado
    patch_tornado()
    USE_JADE = True
except:  # pylint: disable=W0702
    USE_JADE = False

import os.path
import os
import json
import uuid
import multiprocessing
import datetime
import time
import errno

import yandex_tank_api.common as common

TRANSFER_SIZE_LIMIT = 128 * 1024
DEFAULT_HEARTBEAT_TIMEOUT = 600


class APIHandler(tornado.web.RequestHandler):  # pylint: disable=R0904

    """
    Parent class for API handlers
    """

    def initialize(self, server):  # pylint: disable=W0221
        """
        sessions
            dict: session_id->session_status
        """
        # pylint: disable=W0201
        server.read_status_updates()
        self.srv = server

    def reply_json(self, status_code, reply):
        """
        Reply with a json and a specified code
        """
        if status_code != 418:
            self.set_status(status_code)
        else:
            self.set_status(status_code, 'I\'m a teapot!')
        self.set_header('Content-Type', 'application/json')
        reply_str = json.dumps(reply, indent=4)
        self.finish(reply_str)

    def reply_reason(self, code, reason):
        return self.reply_json(code, {'reason': reason})

    def write_error(self, status_code, **kwargs):
        if self.settings.get("debug"):
            tornado.web.RequestHandler(self, status_code, **kwargs)
            return

        self.set_header('Content-Type', 'application/json')
        if 'exc_info' in kwargs and status_code >= 400 and status_code < 500:
            self.reply_json(
                status_code, {'reason': str(kwargs['exc_info'][1])})
        else:
            self.reply_json(status_code, {'reason': self._reason})


class RunHandler(APIHandler):  # pylint: disable=R0904

    """
    Handles POST /run and get /run
    """

    def post(self):

        offered_test_id = self.get_argument("test", datetime.datetime.now().strftime('%Y%m%d%H%M%S'))
        breakpoint = self.get_argument("break", "finished")
        hb_timeout = self.get_argument("heartbeat", None)

        config = self.request.body

        # 503 if any running session exists
        if self.srv.running_id is not None:
            reply = {'reason': 'Another session is already running.'}
            reply.update(self.srv.running_status)
            self.reply_json(503, reply)
            return

        # 400 if invalid breakpoint
        if not common.is_valid_break(breakpoint):
            self.reply_json(
                400,
                {'reason': 'Specified break is not a valid test stage name.',
                 'hint': {'breakpoints': common.get_valid_breaks()}}
            )
            return
        try:
            session_id = self.srv.create_session_dir(offered_test_id)
        except RuntimeError as err:
            self.reply_reason(500, str(err))
            return

        # Remember that such session exists
        self.srv.set_session_status(
            session_id,
            {'status': 'starting', 'break': breakpoint}
        )
        # Post run command to manager queue
        self.srv.cmd({'session': session_id,
                      'cmd': 'run',
                      'break': breakpoint,
                      'config': config})

        self.srv.heartbeat(session_id, hb_timeout)
        self.reply_json(200, {"session": session_id})

    def get(self):
        breakpoint = self.get_argument("break", "finished")
        session_id = self.get_argument("session")
        hb_timeout = self.get_argument("heartbeat", None)

        self.set_header("Content-type", "application/json")

        # 400 if invalid breakpoint
        if not common.is_valid_break(breakpoint):
            self.reply_json(
                400, {
                    'reason': 'Specified break is not a valid test stage name.',
                    'hint':
                    {'breakpoints': common.get_valid_breaks()}
                })
            return

        # 404 if no such session
        try:
            status_dict = self.srv.status(session_id)
        except KeyError:
            self.reply_reason(404, 'No session with this ID.')
            return

	if session_id!=self.srv.running_id:
            self.reply_reason(418,
                "I'm a teapot! Can't set break for session that's not running!")
            return

        # 418 if in higher state
        if common.is_A_earlier_than_B(breakpoint, status_dict['break']):
            reply = {'reason': 'I am a teapot! I know nothing of time-travel!',
                     'hint': {'breakpoints': common.get_valid_breaks()}}
            reply.update(status_dict)
            self.reply_json(418, reply)
            return

        # Post run command to manager queue
        self.srv.cmd({
            'session': session_id,
            'cmd': 'run',
            'break': breakpoint})

        self.srv.heartbeat(session_id, hb_timeout)
        self.reply_reason(200, "Will try to set break before " + breakpoint)


class StopHandler(APIHandler):  # pylint: disable=R0904

    """
    Handles GET /stop
    """

    def get(self):
        session_id = self.get_argument("session")

        try:
            _ = self.srv.status(session_id)
        except KeyError:
            self.reply_reason(404, 'No session with this ID.')
            return
        if self.srv.running_id == session_id:
            self.srv.cmd({'cmd': 'stop', 'session': session_id})
            self.reply_reason(200, 'Will try to stop tank process.')
            return
        else:
            self.reply_reason(409, 'This session is already stopped.')
            return


class StatusHandler(APIHandler):  # pylint: disable=R0904

    """
    Handle GET /status?
    """

    def get(self):
        session_id = self.get_argument("session", default=None)
        if session_id:
            try:
                status = self.srv.status(session_id)
            except KeyError:
                self.reply_reason(404, 'No session with this ID.')
            self.srv.heartbeat(session_id)
            self.reply_json(200, status)
        else:
            self.reply_json(200, self.srv.all_sessions)


class UploadHandler(APIHandler):  # pylint: disable=R0904

    """
    Handles POST /upload
    """

    def post(self):
        session_id = self.get_argument("session")
        if session_id != self.srv.running_id:
            self.reply_reason(404, 'Specified session is not running')
            return

        filename = self.get_argument("filename")
        contents = self.request.body

        filepath = self.srv.session_file(session_id, filename)
        tmp_path = filepath + str(uuid.uuid4())

        with open(tmp_path, 'wb') as upload_file:
            upload_file.write(contents)
        os.rename(tmp_path, filepath)

        self.srv.heartbeat(session_id)
        self.reply_reason(200, 'File uploaded')


class ArtifactHandler(APIHandler):  # pylint: disable=R0904

    """
    Handle GET /atrifact?
    """

    def get(self):
        session_id = self.get_argument("session")

        filename = self.get_argument("filename", None)
        maxsize = self.get_argument("maxsize", None)

        # look for test directory
        if not os.path.exists(self.srv.session_dir(session_id)):
            self.reply_reason(404, 'No session with this ID found')
            return

        # look for status.json (any test that went past lock stage should have
        # it)
        if self.srv.is_empty_session(session_id):
            self.reply_reason(404, 'Test was not performed, no artifacts.')
            return

        if not filename:
            basepath = self.srv.session_dir(session_id)
            onlyfiles = [
                f for f in os.listdir(basepath)
                if os.path.isfile(os.path.join(basepath, f))
            ]
            self.reply_json(200, onlyfiles)
            return

        filepath = self.srv.session_file(session_id, filename)
        if not os.path.exists(filepath):
            self.reply_reason(404, 'No such file in test artifacts')
            return
        file_size = os.stat(filepath).st_size

        if maxsize is not None and file_size > maxsize:
            self.reply_json(409,
                            {'reason': "File does not fit into the size limit specified by the client.",
                             'filesize': file_size})
            return

        if file_size > TRANSFER_SIZE_LIMIT:
            try:
                cur_stage = self.srv.running_status['current_stage']
            except KeyError:
                pass
            else:
                if common.is_A_earlier_than_B(cur_stage, 'postprocess'):
                    self.reply_json(503, {
                        'reason':
                        'File is too large and a session is running',
                        'running_session': self.srv.running_id,
                        'filesize': file_size,
                        'limit': TRANSFER_SIZE_LIMIT
                    })
                    return
        self.set_header("Content-type", "application/octet-stream")
        with open(filepath, 'rb') as artifact_file:
            while True:
                data = artifact_file.read(TRANSFER_SIZE_LIMIT)
                if not data:
                    break
                self.write(data)
                self.flush()
        self.finish()
        self.srv.heartbeat(session_id)


class StaticHandler(tornado.web.RequestHandler):  # pylint: disable=R0904

    """
    Handle /manager.html
    """

    def initialize(self, templ):  # pylint: disable=W0221
        self.template = templ  # pylint: disable=W0201

    def get(self):
        self.render(self.template)


class ApiServer(object):

    """ API server class"""

    def __init__(self, in_queue, out_queue, working_dir, debug=False):
        self._in_queue = in_queue
        self._out_queue = out_queue
        self._working_dir = working_dir
        self._running_id = None
        self._sessions = {}
        self._hb_deadline = None
        self._hb_timeout = DEFAULT_HEARTBEAT_TIMEOUT

        handler_params = dict(server=self)

        handlers = [
            (r"/run", RunHandler, handler_params),
            (r"/stop", StopHandler, handler_params),
            (r"/status", StatusHandler, handler_params),
            (r"/artifact", ArtifactHandler, handler_params),
            (r"/upload", UploadHandler, handler_params)

        ]

        if USE_JADE:
            handlers.append(
                (r"/manager\.html$", StaticHandler,
                 {"template": "manager.jade"})
            )

        self.app = tornado.web.Application(
            handlers,
            template_path=os.path.join(os.path.dirname(__file__), "templates"),
            static_path=os.path.join(os.path.dirname(__file__), "static"),
            debug=debug,
        )

    def read_status_updates(self):
        """Read status messages from manager"""
        try:
            while True:
                message = self._in_queue.get_nowait()
                session_id = message.get('session')
                del message['session']
                self.set_session_status(session_id, message)
        except multiprocessing.queues.Empty:
            pass

    def check(self):
	"""Read status messages from manager and check heartbeat"""
        self.read_status_updates()	

        if self._running_id and self._hb_deadline is not None\
                and time.time() > self._hb_deadline:
            
            self.cmd({'cmd': 'run', 'session': self._running_id, 'break':'finished'})
            self.cmd({'cmd': 'stop', 'session': self._running_id})

    def set_session_status(self, session_id, new_status):
        """Remember session status and change running_id"""

        if new_status['status'] in ['success', 'failed']:
            if self._running_id == session_id:
                self._running_id = None
        else:
            self._running_id = session_id

        self._sessions[session_id] = new_status

    def heartbeat(self, session_id, new_timeout=None):
        """
        Set new heartbeat timeout (if sepcified)
        and reset heartbeat deadline
        """
        if new_timeout is not None:
            self._hb_timeout = new_timeout
        if session_id == self._running_id and self._running_id is not None:
            self._hb_deadline = time.time() + self._hb_timeout

    def session_dir(self, session_id):
        """Return working directory for given session id"""
        return os.path.join(self._working_dir, session_id)

    def session_file(self, session_id, filename):
        """Return file path for given session id"""
        return os.path.join(self._working_dir, session_id, filename)

    def create_session_dir(self, offered_id):
        """
        Returns generated session id
        Should only be used if no tests are running
        """
        if not offered_id:
            offered_id = datetime.datetime.now().strftime("%Y%m%d%H%M%S")
        # This should use one or two attempts in typical cases
        for n_attempt in xrange(10000000000):
            session_id = "%s_%010d" % (offered_id, n_attempt)
            session_dir = self.session_dir(session_id)
            try:
                os.makedirs(session_dir)
            except OSError as err:
                if err.errno != errno.EEXIST:
                    raise RuntimeError("Failed to create session directory")
            if self.is_empty_session(session_id):
                return session_id
            n_attempt += 1
        raise RuntimeError("Failed to generate session id")

    def is_empty_session(self, session_id):
        """Return true if the session did not get past the lock stage"""
        return not os.path.exists(self.session_file(session_id, 'status.json'))

    def cmd(self, message):
        """Put commad into manager queue"""
        self._out_queue.put(message)

    @property
    def all_sessions(self):
        """Get session status by ID, can raise KeyError"""
        return self._sessions

    def status(self, session_id):
        """Get session status by ID, can raise KeyError"""
        return self._sessions[session_id]

    @property
    def running_id(self):
        """Return ID of running session"""
        return self._running_id

    @property
    def running_status(self):
        """Return status of running session , can raise KeyError"""
        return self.status(self._running_id)

    def serve(self):
        """
        Run tornado ioloop
        """
        self.app.listen(8888)
        ioloop = tornado.ioloop.IOLoop.instance()
        update_cb = tornado.ioloop.PeriodicCallback(
            self.check, 300, ioloop)
        update_cb.start()
        ioloop.start()


def main(webserver_queue, manager_queue, test_directory, debug):
    """Target for webserver process.
    The only function ever used by the Manager.

    webserver_queue
        Read statuses from Manager here.

    manager_queue
        Write commands for Manager there.

    test_directory
        Directory where tests are

    """
    ApiServer(webserver_queue, manager_queue, test_directory, debug).serve()
